atef/widgets/happi.py

```diff
@@ -0,0 +1,470 @@
+        self.proxy_model.setFilterKeyColumn(-1)
+        self.proxy_model.setDynamicSortFilter(True)
+        self.proxy_model.setSourceModel(self.model)
+        self.table_view.setModel(self.proxy_model)
+
+        self.table_view.setContextMenuPolicy(QtCore.Qt.CustomContextMenu)
+        self.table_view.customContextMenuRequested.connect(self._table_context_menu)
+
+    def _table_context_menu(self, pos: QtCore.QPoint) -> None:
+        """Context menu when the key/value table is right-clicked."""
+        self.menu = QtWidgets.QMenu(self)
+        index: QtCore.QModelIndex = self.table_view.indexAt(pos)
+        if index is not None:
+            def copy(*_):
+                copy_to_clipboard(index.data())
+
+            copy_action = self.menu.addAction(f"&Copy: {index.data()}")
+            copy_action.triggered.connect(copy)
+
+        self.menu.exec_(self.table_view.mapToGlobal(pos))
+
+    def _update_metadata(self):
+        """Update the metadata based on ``self.item_name`` using the configured client."""
+        if self.client is None or self.item_name is None:
+            return
+
+        try:
+            self.item = self.client[self.item_name]
+        except KeyError:
+            self.item = None
+
+        metadata = dict(self.item or {})
+        self._metadata = metadata
+        self.updated_metadata.emit(self.item_name, metadata)
+        self.model.clear()
+        if self.item is None:
+            self.label_title.setText("")
+            return
+
+        self.label_title.setText(metadata["name"])
+        self.model.setHorizontalHeaderLabels(["Key", "Value"])
+        skip_keys = {"_id", "name"}
+        for key, value in sorted(metadata.items()):
+            if key in skip_keys:
+                continue
+
+            key_item = QtGui.QStandardItem(str(key))
+            value_item = QtGui.QStandardItem(str(value))
+            key_item.setFlags(key_item.flags() & ~QtCore.Qt.ItemIsEditable)
+            value_item.setFlags(value_item.flags() & ~QtCore.Qt.ItemIsEditable)
+            self.model.appendRow([key_item, value_item])
+
+    @property
+    def client(self) -> Optional[happi.Client]:
+        """The client to use for search."""
+        return self._client
+
+    @client.setter
+    def client(self, client: Optional[happi.Client]):
+        self._client = client
+        self._update_metadata()
+
+    @property
+    def item_name(self) -> Optional[str]:
+        """The item name to search for metadata."""
+        return self._item_name
+
+    @item_name.setter
+    def item_name(self, item_name: Optional[str]):
+        self._item_name = item_name
+        self._update_metadata()
+
+    @property
+    def metadata(self) -> Dict[str, Any]:
+        """The current happi item metadata, as a dictionary."""
+        return dict(self._metadata)
+
+
+class HappiDeviceComponentWidget(DesignerDisplay, QWidget):
+    """
+    Happi item (device) search widget + component view and selection.
```

Comment:
For the config gui there must be a way to use this wholesale and show it with a pre-filter on the device list to only include the devices available in the config.

I'm tempted to ignore the problems related to non-uniform signal sets for now.

Reply:
The filter should support regular expressions, so the easiest way would probably be:
reach in to the instantiated dialog's filter textbox and call `setText(regex)`.
It wouldn't cause them to be selected, but I guess it's better than nothing?